File: BillPayments/Core/UiEventBus.swift
final class UiEventBus {
    let events: AsyncStream<UiEvent>
    private let continuation: AsyncStream<UiEvent>.Continuation

    init() {
        (events, continuation) = AsyncStream.makeStream(bufferingPolicy: .unbounded)
    }

    func emit(_ event: UiEvent) {
        continuation.yield(event)
    }
}

File: app/src/main/java/com/billpayments/core/event/UiEventBus.kt
package com.billpayments.core.event

import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.receiveAsFlow

@Singleton
class UiEventBus @Inject constructor() {
    private val channel = Channel<UiEvent>(Channel.BUFFERED)
    val events: Flow<UiEvent> = channel.receiveAsFlow()

    fun emit(event: UiEvent) {
        channel.trySend(event)
    }
}
